### ISSUE TITLE
stats JSON endpoint fails to fully load

### ISSUE THREAD

**Issue by debtsandbooze** [haproxy/haproxy#1873]
### Detailed Description of the Problem

For HAProxy versions after 2.5.x; Querying the JSON stats endpoint under moderate backend server load/connections, we are seeing the json return partial results and hang. For one of our monitoring solutions we simply poll the stats endpoint every minute. I was able to reproduce in versions 2.6.x to 2.7.x. Additionally the CSV endpoint does not have any issues and fully loads after each query. Example query endpoint we are hitting: http://hostname:8084/stats;json

### Expected Behavior

Under no load/connections we can query the JSON endpoint all day without issue.

### Steps to Reproduce the Behavior

- Run a version of HAProxy 2.6.x or higher
- generate load to your backends
- Attempt to curl or load your JSON stats URL

### Do you have an idea how to solve the issue?

Reverting to haproxy-2.5 fixes the issue

### What is your configuration?

```haproxy
# ---------------------------------------------------------------------
# Generated by constellation (config generator)
# ADC: '${adc}'
# Template: '${template}'
# ---------------------------------------------------------------------

# ---------------------------------------------------------------------
# Global settings
# ---------------------------------------------------------------------

global
  chroot /var/lib/haproxy
  user haproxy
  group haproxy
  maxconn 10000
  pidfile /var/run/haproxy.pid
  stats socket /var/lib/haproxy/stats
  log stdout format raw sample 1:1000 local0 debug

# ---------------------------------------------------------------------
# common defaults that all the 'listen' and 'backend' sections will
# use if not designated in their block
# ---------------------------------------------------------------------
defaults
  mode http
  log global
  option httplog
  option redispatch
  option dontlognull
  option http-server-close
  timeout http-request 10s
  timeout check 10s
  timeout connect 10s
  timeout client 1m
  timeout queue 1m
  timeout server 1m
  timeout http-keep-alive 10s
  retries 3

# option forwardfor except 127.0.0.0/8
# ---------------------------------------------------------------------
# example how to define user and enable Data Plane API on tcp/5555
# more information: https://github.com/haproxytech/dataplaneapi and
# https://www.haproxy.com/documentation/hapee/2-0r1/configuration/dataplaneapi/
# ---------------------------------------------------------------------
userlist haproxy-dataplaneapi
  user admin insecure-password mypassword

# Custom Error Pages
http-errors all-brands
  errorfile 403 /etc/haproxy/errors/403-all-brands.http

frontend fe-dataplaneapi
  bind 0.0.0.0:5554
  acl valid_method method GET OPTION HEAD
  http-request deny if !valid_method
  default_backend be-dataplaneapi

# ---------------------------------------------------------------------
# main frontend which proxys to the backends
# ---------------------------------------------------------------------
frontend main-ssl
  maxconn 100000
  bind 0.0.0.0:80
  bind 0.0.0.0:443 ssl crt /usr/local/etc/haproxy/certs/${certificate} ssl-min-ver TLSv1.2 alpn h2,http/1.1
  option httpslog
  log-format "{\"client_ip\": \"%ci\", \"client_port\":\"%cp\", \"date_time\": \"%tr\", \"frontend_name_transport\": \"%ft\",\"backend_name\": \"%b\", \"server_name\": \"%s\", \"time_to_receive\": \"%TR\", \"Tw\": \"%Tw\", \"Tc\": \"%Tc\", \"resp_time\": \"%Tr\",\"active_time\": \"%Ta\", \"status_code\": \"%ST\", \"bytes_read\": \"%B\", \"captured_request_cookie\": \"%CC\", \"captured_response_cookie\": \"%CS\", \"termination_state_with_cookie\": \"%tsc\",\"actconn\": \"%ac\", \"feconn\": \"%fc\", \"beconn\": \"%bc\", \"srv_conn\": \"%sc\", \"retries\": \"%rc\", \"srv_queue\": \"%sq\", \"backend_queue\": \"%bq\", \"captured_request_headers\": \"%hr\",\"captured_response_headers\": \"%hs\", \"http_request\": \"%r\", \"fc_err\":\"%[fc_err]\", \"ssl_fc_err\": \"%[ssl_fc_err]\", \"ssl_c_err\": \"%[ssl_c_err]\", \"ssl_c_ca_err\": \"%[ssl_c_ca_err]\", \"ssl_fc_is_resumed\": \"%[ssl_fc_is_resumed]\", \"ssl_fc_sni\": \"%[ssl_fc_sni]\", \"ssl_version\" : \"%sslv\", \"ssl_ciphers\": \"%sslc\" }"
  option forwardfor
  timeout http-request 60s

  # Forward Original http protocol scheme
  http-request set-header X-Forwarded-Proto https if { ssl_fc }
  http-request set-header X-Forwarded-Proto http if !{ ssl_fc }

  # Cutstom Errors
  errorfiles all-brands

  # Deny apache server-status page from frontend
  http-request deny if { path_beg -i /server-status }

  # Use True-Client-IP as the source IP address
  http-request set-src hdr(True-Client-IP)

  # Modsecurity WAF Deny
  http-request deny if { req.hdr_cnt(content-length) gt 1 }
  http-response set-header X-SN-WAF-Code %[var(txn.modsec.code)]

  # Default Backend
  default_backend blah


# ---------------------------------------------------------------------
# round robin balancing between the various backends
# ---------------------------------------------------------------------
backend be-dataplaneapi
  server dataplane 127.0.0.1:5555

backend blah
  balance roundrobin
  option accept-invalid-http-response

  # Health Check
  option httpchk
  http-check send meth GET uri /x ver HTTP/1.1 hdr Host www.hostname.com hdr X-Monitor haproxyc hdr User-Agent hostname-haproxy/1.0
  http-check expect rstring OK

  # Enable Mod Security
  filter spoe engine modsecurity config /etc/haproxy/spoe-modsecurity.conf
  http-request deny if { var(txn.modsec.code) -m int gt 0 }

  http-request add-header SSL NO if !{ ssl_fc }
  http-request add-header SSL YES if { ssl_fc }

  http-response set-header X-SN-WAF-Code %[var(txn.modsec.code)]

  # Define Backends in adc.json
  $(printf '%s\n  ' "${backends[@]}")

backend spoe-modsecurity
  mode tcp
  server modsec-spoa modsecurity:12345 check

# ---------------------------------------------------------------------
# random comment
# ---------------------------------------------------------------------
listen stats
  bind 0.0.0.0:8084
  http-request use-service prometheus-exporter if { path /metrics }
  stats enable
  stats refresh 30s
  stats uri /stats

# ---------------------------------------------------------------------
# running data plane api using HAProxy Process manager
# ---------------------------------------------------------------------
program api
  command /usr/local/bin/dataplaneapi --host 127.0.0.1 --port 5555 --haproxy-bin /usr/local/sbin/haproxy --config-file /etc/haproxy/haproxy.cfg --reload-cmd "kill -SIGUSR2 1" --restart-cmd "kill -SIGUSR2 1" --reload-delay 5 --userlist haproxy-dataplaneapi --log-to=stdout --log-level=debug --log-format=text
  no option start-on-reload
```

### Output of `haproxy -vv`

```plain
/ # haproxy -vv
HAProxy version 2.6.4-2a2078c 2022/08/22 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.4.html
Running on: Linux 5.4.212-1.el7.elrepo.x86_64 #1 SMP Sat Sep 3 13:19:51 EDT 2022 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1q  5 Jul 2022
Running on OpenSSL version : OpenSSL 1.1.1q  5 Jul 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.3.1 20211027

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by capflam** [comment]
Have you tried the last 2.7 release (2.7-dev6) or the 2.6.5 ? If possible, it could be useful to get a network capture on the port 8084. In addition, what is your 2.5 version ?

**Comment by debtsandbooze** [comment]
We've tried the Dev(2.7) build and 2.6.5, both appear to hang with the JSON stats endpoint when testing upwards of 30,000 requests per second. Currently running 2.5.8 without issues. Not seeing anything out of the norm in a pcap. 

/ # haproxy -v
HAProxy version 2.5.8-0cbd0f6 2022/07/25 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2023.
Known bugs: http://www.haproxy.org/bugs/bugs-2.5.8.html
Running on: Linux 5.4.212-1.el7.elrepo.x86_64 #1 SMP Sat Sep 3 13:19:51 EDT 2022 x86_64

**Comment by capflam** [comment]
Sorry, I missed the update. Thanks for the info. you said there is nothing special in the pcap. The request hangs with no response or the response is partially transferred ? In the last case, what is the amount of data transferred ? If it is possible you can share your pcap in private. It may ease the analysis. But I'll truly understand if it is not possible.

**Comment by debtsandbooze** [comment]
The response is partially sent and the connection remains open. Unfortunately my security team would frown upon sending the pcap. Too much to even consider sanitizing. It was pretty easy to reproduce with an AB load test. Just need to ramp up connections.

**Comment by capflam** [comment]
It probably depends on your configuration or you environment because I'm unable to reproduce the bug. Note it is not related to json output itself. It is most probably a condition not properly checked (a buffer full, an early connection shutdown, ...) blocking the data transfer. How many backends and server per backends do you have ?

**Comment by wtarreau** [comment]
Since you could try with 2.7-dev, it would be nice to take the latest 2.7-dev and apply this patch on top of it. It will emit on stdout each location in the stats code where we consider a buffer as full and decide to stop sending. It will help us spot where the problem is. Note: you must absolutely not run this in production, and you'll need to run in foreground mode (`-db`) to catch the output. But since you said it's easy to reproduce with your config, I'm assuming this shouldn't be an issue. Thanks!
[stats-show-freespace.diff.txt](https://github.com/haproxy/haproxy/files/9786544/stats-show-freespace.diff.txt)

**Comment by capflam** [comment]
Thanks to @Darlelet, we've finally found the bug. A fix was pushed in 2.8-DEV and will most probably be shipped with the 2.7.1.

**Comment by debtsandbooze** [comment]
Thank you @Darlelet. I upgraded a node to 2.7.2 and we started encountering the same bug as initially reported. stats:json seems to hang while having around 20 concurrent connections to it.

/ # haproxy -v
HAProxy version 2.7.2-7e295dd 2023/01/20 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2024.
Known bugs: http://www.haproxy.org/bugs/bugs-2.7.2.html
Running on: Linux 5.4.212-1.el7.elrepo.x86_64 #1 SMP Sat Sep 3 13:19:51 EDT 2022 x86_64

**Comment by Darlelet** [comment]
Thank you for testing and reporting back to us.
You're right, I'm able to reproduce, seems like https://github.com/haproxy/haproxy/commit/a8b7684319a48110afb61d50f004566886a6baf1 is not enough unfortunately

**Comment by debtsandbooze** [comment]
Should I open a new issue or can we reopen and run with this one?

**Comment by Darlelet** [comment]
Sorry I missed the update.
FYI: this should be fixed in 2.7.3
